Source/stores.cpp

```diff
@@ -97,47 +97,105 @@ struct STextStruct {
 	}
 };
 
-/** Text lines */
-STextStruct TextLine[STORE_LINES];
-
-/** Whether to render the player's gold amount in the top left */
-bool RenderGold;
-
-/** Does the current panel have a scrollbar */
-bool HasScrollbar;
-/** Remember last scroll position */
-int OldScrollPos;
-/** Scroll position */
-int ScrollPos;
-/** Next scroll position */
-int NextScrollPos;
-/** Previous scroll position */
-int PreviousScrollPos;
-/** Countdown for the push state of the scroll up button */
-int8_t CountdownScrollUp;
-/** Countdown for the push state of the scroll down button */
-int8_t CountdownScrollDown;
-
-/** Remember current store while displaying a dialog */
-TalkID OldActiveStore;
-
-/** Temporary item used to hold the item being traded */
-Item TempItem;
-
-/** Maps from towner IDs to NPC names. */
-const char *const TownerNames[] = {
-	N_("Griswold"),
-	N_("Pepin"),
-	"",
-	N_("Ogden"),
-	N_("Cain"),
-	N_("Farnham"),
-	N_("Adria"),
-	N_("Gillian"),
-	N_("Wirt"),
+std::array<STextStruct, NumStoreLines> TextLine; // Text lines
+
+bool RenderGold;            // Whether to render the player's gold amount in the top left
+int OldScrollPos;           // Remember last scroll position
+int ScrollPos;              // Scroll position
+int NextScrollPos;          // Next scroll position
+int PreviousScrollPos;      // Previous scroll position
+int8_t CountdownScrollUp;   // Countdown for the push state of the scroll up button
+int8_t CountdownScrollDown; // Countdown for the push state of the scroll down button
+
+TalkID OldActiveStore; // Remember current store while displaying a dialog
+
+Item TempItem; // Temporary item used to hold the item being traded
+
+std::vector<std::pair<int, TalkID>> LineActionMappings;
+int CurrentMenuDrawLine;
+
+const std::string SmithMenuHeader = "Welcome to the\n\nBlacksmith's shop";
+
+const StoreMenuOption SmithMenuOptions[] = {
+	{ TalkID::Gossip, fmt::format("Talk to {:s}", Blacksmith.name) },
+	{ TalkID::BasicBuy, "Buy basic items" },
+	{ TalkID::Buy, "Buy premium items" },
+	{ TalkID::Sell, "Sell items" },
+	{ TalkID::Repair, "Repair items" },
+	{ TalkID::Exit, "Leave the shop" }
+};
+
+const std::string HealerMenuHeader = "Welcome to the\n\nHealer's home";
+
+const StoreMenuOption HealerMenuOptions[] = {

```

Comment:
Having each store statically hard coded doesn't feel like a good direction. It would probably be better to populate a single structure with the relevant store, that would also make it feasible to make custom stores at runtime in mods.